src/python/pants/engine/internals/graph.py

```diff
@@ -559,6 +559,26 @@ def visit(address: Address):
             )
 
 
+@dataclass
+class _DependencyMappings:
+    """A mutable @rule output which stores memoized dependencies."""
+
+    dependencies: dict[Address, tuple[Target, ...]]
+
+
+@dataclass(frozen=True)
+class _DependencyMappingsRequest:
+    expanded_targets: bool
+    include_special_cased_deps: bool
```

Comment:
Slightly off topic:
This PR and #19155 conflict a bit, so I'm thinking about how they could work together. I think `do_traverse_deps_predicate` would have to be included in this request as well.

Reply:
I refactored #19155. A side effect of that refactor is that it no longer conflicts with changes to the DependencyMappings rule.